nil dereference panic is occured, when the structure has same type fields?

Inputting the code something like this, nil dereference panic is occurred.

```go
package main

import (
	"os"

	"github.com/goccy/go-json"
)

type S0 struct {
	X S1 `json:"x"`
	Y S1 `json:"y"`
}

type S1 struct {
	Name string `json:"name"`
}

func main() {
	s := S0{X: S1{Name: "foo"}, Y: S1{Name: "bar"}}
	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	encoder.Encode(&s)
}
```

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1 pc=0x10b164a]
```

<details>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1 pc=0x10b164a]

goroutine 1 [running]:
github.com/goccy/go-json.(*Encoder).encodeEscapedString(0xc0000bc000, 0x1, 0xc000099f98)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_string.go:226 +0x7a
github.com/goccy/go-json.(*Encoder).encodeString(0xc0000bc000, 0x1, 0xc000099f98)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode.go:271 +0x45
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6700, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:1002 +0xa38f
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6680, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6600, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6580, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6500, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6480, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6400, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6380, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6300, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6280, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6200, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6180, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6100, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6080, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c6000, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1f80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1f00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1e80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1e00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1d80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1d00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1c80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1c00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1b80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1b00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1a80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1a00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1980, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1900, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1880, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1800, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1780, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1700, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1680, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1600, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1580, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1500, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1480, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1400, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1380, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1300, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1280, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1200, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1180, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1100, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1080, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c1000, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0f80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0f00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0e80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0e00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0d80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0d00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0c80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0c00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0b80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0b00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0a80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0a00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0980, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0900, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0880, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0800, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0780, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0700, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0680, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0600, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0580, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0500, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0480, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0400, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0380, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0300, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0280, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0200, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0180, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0100, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0080, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000c0000, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abf80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abf00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abe80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abe00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abd80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abd00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abc80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abc00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abb80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000abb00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000aba80, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000aba00, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000ab980, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000ab900, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000ab880, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000ab800, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
github.com/goccy/go-json.(*Encoder).run(0xc0000bc000, 0xc0000ab780, 0x4, 0x0)
        ~/go/pkg/mod/github.com/goccy/go-json@v0.1.3/encode_vm.go:465 +0x11cec
```

</details>

--- a/encode_compile.go
+++ b/encode_compile.go
@@ -749,7 +749,7 @@ func (e *Encoder) optimizeStructField(op opType, isOmitEmpty, withIndent bool) o
 func (e *Encoder) compileStruct(typ *rtype, root, withIndent bool) (*opcode, error) {
 	typeptr := uintptr(unsafe.Pointer(typ))
 	if withIndent {
-		if compiled, exists := e.structTypeToCompiledCode[typeptr]; exists {
+		if compiled, exists := e.structTypeToCompiledIndentCode[typeptr]; exists {
 			return (*opcode)(unsafe.Pointer(&recursiveCode{
 				opcodeHeader: &opcodeHeader{
 					op:     opStructFieldRecursive,
@@ -761,7 +761,7 @@ func (e *Encoder) compileStruct(typ *rtype, root, withIndent bool) (*opcode, err
 			})), nil
 		}
 	} else {
-		if compiled, exists := e.structTypeToCompiledIndentCode[typeptr]; exists {
+		if compiled, exists := e.structTypeToCompiledCode[typeptr]; exists {
 			return (*opcode)(unsafe.Pointer(&recursiveCode{
 				opcodeHeader: &opcodeHeader{
 					op:     opStructFieldRecursive,
@@ -775,9 +775,9 @@ func (e *Encoder) compileStruct(typ *rtype, root, withIndent bool) (*opcode, err
 	}
 	compiled := &compiledCode{}
 	if withIndent {
-		e.structTypeToCompiledCode[typeptr] = compiled
-	} else {
 		e.structTypeToCompiledIndentCode[typeptr] = compiled
+	} else {
+		e.structTypeToCompiledCode[typeptr] = compiled
 	}
 	// header => code => structField => code => end
 	//                        ^          |

--- a/encode_opcode.go
+++ b/encode_opcode.go
@@ -1416,3 +1416,134 @@ func (c *recursiveCode) copy(codeMap map[uintptr]*opcode) *opcode {
 	}
 	return code
 }
+
+func newRecursiveCode(recursive *recursiveCode) *opcode {
+	code := copyOpcode(recursive.jmp.code)
+	head := (*structFieldCode)(unsafe.Pointer(code))
+	head.end.next = newEndOp(0)
+	code.ptr = recursive.ptr
+
+	switch code.op {
+	case opStructFieldPtrHead:
+		code.op = opStructFieldHead
+	case opStructFieldPtrHeadInt:
+		code.op = opStructFieldHeadInt
+	case opStructFieldPtrHeadInt8:
+		code.op = opStructFieldHeadInt8
+	case opStructFieldPtrHeadInt16:
+		code.op = opStructFieldHeadInt16
+	case opStructFieldPtrHeadInt32:
+		code.op = opStructFieldHeadInt32
+	case opStructFieldPtrHeadInt64:
+		code.op = opStructFieldHeadInt64
+	case opStructFieldPtrHeadUint:
+		code.op = opStructFieldHeadUint
+	case opStructFieldPtrHeadUint8:
+		code.op = opStructFieldHeadUint8
+	case opStructFieldPtrHeadUint16:
+		code.op = opStructFieldHeadUint16
+	case opStructFieldPtrHeadUint32:
+		code.op = opStructFieldHeadUint32
+	case opStructFieldPtrHeadUint64:
+		code.op = opStructFieldHeadUint64
+	case opStructFieldPtrHeadFloat32:
+		code.op = opStructFieldHeadFloat32
+	case opStructFieldPtrHeadFloat64:
+		code.op = opStructFieldHeadFloat64
+	case opStructFieldPtrHeadString:
+		code.op = opStructFieldHeadString
+	case opStructFieldPtrHeadBool:
+		code.op = opStructFieldHeadBool
+	case opStructFieldPtrHeadIndent:
+		code.op = opStructFieldHeadIndent
+	case opStructFieldPtrHeadIntIndent:
+		code.op = opStructFieldHeadIntIndent
+	case opStructFieldPtrHeadInt8Indent:
+		code.op = opStructFieldHeadInt8Indent
+	case opStructFieldPtrHeadInt16Indent:
+		code.op = opStructFieldHeadInt16Indent
+	case opStructFieldPtrHeadInt32Indent:
+		code.op = opStructFieldHeadInt32Indent
+	case opStructFieldPtrHeadInt64Indent:
+		code.op = opStructFieldHeadInt64Indent
+	case opStructFieldPtrHeadUintIndent:
+		code.op = opStructFieldHeadUintIndent
+	case opStructFieldPtrHeadUint8Indent:
+		code.op = opStructFieldHeadUint8Indent
+	case opStructFieldPtrHeadUint16Indent:
+		code.op = opStructFieldHeadUint16Indent
+	case opStructFieldPtrHeadUint32Indent:
+		code.op = opStructFieldHeadUint32Indent
+	case opStructFieldPtrHeadUint64Indent:
+		code.op = opStructFieldHeadUint64Indent
+	case opStructFieldPtrHeadFloat32Indent:
+		code.op = opStructFieldHeadFloat32Indent
+	case opStructFieldPtrHeadFloat64Indent:
+		code.op = opStructFieldHeadFloat64Indent
+	case opStructFieldPtrHeadStringIndent:
+		code.op = opStructFieldHeadStringIndent
+	case opStructFieldPtrHeadBoolIndent:
+		code.op = opStructFieldHeadBoolIndent
+	case opStructFieldPtrHeadOmitEmpty:
+		code.op = opStructFieldHeadOmitEmpty
+	case opStructFieldPtrHeadIntOmitEmpty:
+		code.op = opStructFieldHeadIntOmitEmpty
+	case opStructFieldPtrHeadInt8OmitEmpty:
+		code.op = opStructFieldHeadInt8OmitEmpty
+	case opStructFieldPtrHeadInt16OmitEmpty:
+		code.op = opStructFieldHeadInt16OmitEmpty
+	case opStructFieldPtrHeadInt32OmitEmpty:
+		code.op = opStructFieldHeadInt32OmitEmpty
+	case opStructFieldPtrHeadInt64OmitEmpty:
+		code.op = opStructFieldHeadInt64OmitEmpty
+	case opStructFieldPtrHeadUintOmitEmpty:
+		code.op = opStructFieldHeadUintOmitEmpty
+	case opStructFieldPtrHeadUint8OmitEmpty:
+		code.op = opStructFieldHeadUint8OmitEmpty
+	case opStructFieldPtrHeadUint16OmitEmpty:
+		code.op = opStructFieldHeadUint16OmitEmpty
+	case opStructFieldPtrHeadUint32OmitEmpty:
+		code.op = opStructFieldHeadUint32OmitEmpty
+	case opStructFieldPtrHeadUint64OmitEmpty:
+		code.op = opStructFieldHeadUint64OmitEmpty
+	case opStructFieldPtrHeadFloat32OmitEmpty:
+		code.op = opStructFieldHeadFloat32OmitEmpty
+	case opStructFieldPtrHeadFloat64OmitEmpty:
+		code.op = opStructFieldHeadFloat64OmitEmpty
+	case opStructFieldPtrHeadStringOmitEmpty:
+		code.op = opStructFieldHeadStringOmitEmpty
+	case opStructFieldPtrHeadBoolOmitEmpty:
+		code.op = opStructFieldHeadBoolOmitEmpty
+	case opStructFieldPtrHeadOmitEmptyIndent:
+		code.op = opStructFieldHeadOmitEmptyIndent
+	case opStructFieldPtrHeadIntOmitEmptyIndent:
+		code.op = opStructFieldHeadIntOmitEmptyIndent
+	case opStructFieldPtrHeadInt8OmitEmptyIndent:
+		code.op = opStructFieldHeadInt8OmitEmptyIndent
+	case opStructFieldPtrHeadInt16OmitEmptyIndent:
+		code.op = opStructFieldHeadInt16OmitEmptyIndent
+	case opStructFieldPtrHeadInt32OmitEmptyIndent:
+		code.op = opStructFieldHeadInt32OmitEmptyIndent
+	case opStructFieldPtrHeadInt64OmitEmptyIndent:
+		code.op = opStructFieldHeadInt64OmitEmptyIndent
+	case opStructFieldPtrHeadUintOmitEmptyIndent:
+		code.op = opStructFieldHeadUintOmitEmptyIndent
+	case opStructFieldPtrHeadUint8OmitEmptyIndent:
+		code.op = opStructFieldHeadUint8OmitEmptyIndent
+	case opStructFieldPtrHeadUint16OmitEmptyIndent:
+		code.op = opStructFieldHeadUint16OmitEmptyIndent
+	case opStructFieldPtrHeadUint32OmitEmptyIndent:
+		code.op = opStructFieldHeadUint32OmitEmptyIndent
+	case opStructFieldPtrHeadUint64OmitEmptyIndent:
+		code.op = opStructFieldHeadUint64OmitEmptyIndent
+	case opStructFieldPtrHeadFloat32OmitEmptyIndent:
+		code.op = opStructFieldHeadFloat32OmitEmptyIndent
+	case opStructFieldPtrHeadFloat64OmitEmptyIndent:
+		code.op = opStructFieldHeadFloat64OmitEmptyIndent
+	case opStructFieldPtrHeadStringOmitEmptyIndent:
+		code.op = opStructFieldHeadStringOmitEmptyIndent
+	case opStructFieldPtrHeadBoolOmitEmptyIndent:
+		code.op = opStructFieldHeadBoolOmitEmptyIndent
+	}
+	return code
+}

--- a/encode_test.go
+++ b/encode_test.go
@@ -12,8 +12,10 @@ import (
 type recursiveT struct {
 	A *recursiveT `json:"a,omitempty"`
 	B *recursiveU `json:"b,omitempty"`
-	C string      `json:"c,omitempty"`
+	C *recursiveU `json:"c,omitempty"`
+	D string      `json:"d,omitempty"`
 }
+
 type recursiveU struct {
 	T *recursiveT `json:"t,omitempty"`
 }
@@ -117,13 +119,18 @@ func Test_Marshal(t *testing.T) {
 				A: &recursiveT{
 					B: &recursiveU{
 						T: &recursiveT{
-							C: "hello",
+							D: "hello",
+						},
+					},
+					C: &recursiveU{
+						T: &recursiveT{
+							D: "world",
 						},
 					},
 				},
 			})
 			assertErr(t, err)
-			assertEq(t, "recursive", `{"a":{"b":{"t":{"c":"hello"}}}}`, string(bytes))
+			assertEq(t, "recursive", `{"a":{"b":{"t":{"d":"hello"}},"c":{"t":{"d":"world"}}}}`, string(bytes))
 		})
 		t.Run("omitempty", func(t *testing.T) {
 			type T struct {

--- a/encode_vm.go
+++ b/encode_vm.go
@@ -460,9 +460,7 @@ func (e *Encoder) run(code *opcode) error {
 			code = c.next
 		case opStructFieldRecursive:
 			recursive := code.toRecursiveCode()
-			c := copyOpcode(recursive.jmp.code)
-			c.ptr = recursive.ptr
-			if err := e.run(c); err != nil {
+			if err := e.run(newRecursiveCode(recursive)); err != nil {
 				return err
 			}
 			code = recursive.next

